kuromoji-core/src/main/java/com/atilika/kuromoji/fst/vm/Program.java

```diff
@@ -38,64 +27,56 @@ public Program() {
         this.cacheFirstIsAccept = new boolean[CACHED_CHAR_RANGE];
     }
 
-
-    public byte getOpcode(int pc) {
-        int internalIndex = pc * BYTES_PER_INSTRUCTIONS;
-        return instruction.get(internalIndex);
+    public boolean isAccept(long instruction) {
+        return (instruction & 0x0000008000000000L) != 0;
     }
 
-    public char getArg1(int pc) {
-        int internalIndex = pc * BYTES_PER_INSTRUCTIONS;
-        return instruction.getChar(internalIndex + 1);
+    public boolean isFail(long instruction) {
+        return (instruction & 0x8000000000000000L) != 0;
     }
 
-    public int getArg2(int pc) {
-        int internalIndex = pc * BYTES_PER_INSTRUCTIONS;
-        return instruction.getInt(internalIndex + 1 + 2);
+    public char getLabel(long instruction) {
+        return (char) (instruction & 0x000000000000FFFFL);
     }
 
-    public int getArg3(int pc) {
-        int internalIndex = pc * BYTES_PER_INSTRUCTIONS;
-        return instruction.getInt(internalIndex + 1 + 2 + 4);
+    public int getTargetAddress(long instruction) {
+        return isFail(instruction) ? -1 : (int) ((instruction >> 40) & 0x00000000007FFFFFL);
     }
 
-    public void addInstruction(byte op) {
-        addInstruction(op, ' ', -1, 0);
+    public int getOutput(long instruction) {
+        return (instruction & 0x0000004000000000L) != 0 ? -1 : (int) ((instruction >> 16) & 0x00000000007FFFFFL);
     }
 
-    public void addInstruction(byte op, char label, int targetAddress, int output) {
-        int currentSizePlusOneInstruction = (this.getNumInstructions() + 1) * BYTES_PER_INSTRUCTIONS;
-        if (currentSizePlusOneInstruction >= BYTES_PER_INSTRUCTIONS * numInstructionsAllocated) {
-            doubleBufferSize();
-        }
+    public void addInstruction(boolean fail, boolean accept, char label, int targetAddress, int output) {
+        assert (targetAddress < 0x7FFFFFFF);
+        assert (output < 0x7FFFFFFF);
 
-        instruction.put(op);
-        instruction.putChar(label);
-        instruction.putInt(targetAddress);
-        instruction.putInt(output);
+        // encode the instruction into a 8 byte long
+        // 3 bytes == target address for 23 bits, msb == fail
```

Comment:
I tested all dictionaries and the biggest is 4.4m.. 

> On 1 Oct 2015, at 15:51, Christian Moen <notifications@github.com> wrote:
> 
> In kuromoji-core/src/main/java/com/atilika/kuromoji/fst/vm/Program.java:
> 
> >  
> > -        instruction.put(op);
> > -        instruction.putChar(label);
> > -        instruction.putInt(targetAddress);
> > -        instruction.putInt(output);
> > +        // encode the instruction into a 8 byte long
> > +        // 3 bytes == target address for 23 bits, msb == fail
> I guess the range for the target address will 8,388,608 (2^23) since the high bit is used to indicate a fail?
> 
> That said, can we theoretically even exceed this jump limit? Will it even be possible to jump this far for any FST as per it's construction algorithm? I'm not so sure that's the case... What do you say, Gerry and Fujinuma-san?
> 
> If the target range here turns out to be a problem with a specific FST, I'm guessing building will fail since we validate that we can lookup all keys?
> 
> Thoughts?
> 
> ―
> Reply to this email directly or view it on GitHub.
>